### ISSUE TITLE
HTTP output plugin thinks that could not flush records

### ISSUE THREAD

**Issue by 2solt** [fluent/fluent-bit#352]
Im trying to send logs to logstash (http input), the data arrives correctly but fluent-bit (v0.11.16) thinks it didn't.

fluent-bit.conf:
```
[SERVICE]
    Flush        5
    Daemon       Off
    Log_Level    debug
    Parsers_File parsers.conf

[INPUT]
    Name                tail
    Path                /tmp/test.log
    Tag                 test
    Refresh_Interval    5
    Mem_Buf_Limit       5MB

[OUTPUT]
    Name            http
    Match           *
    Host            192.168.0.89
    Port            5044
    URI             /
    Format          json
```
/tmp/test.log:
```
fluentbit_test
```

fluent-bit log:
```
[2017/08/11 08:52:23] [ info] [engine] started
[2017/08/11 08:52:23] [debug] [in_tail] inotify watch fd=18
[2017/08/11 08:52:23] [debug] [in_tail] scanning path /tmp/test.log
[2017/08/11 08:52:23] [debug] [in_tail] add to scan queue /tmp/test.log, offset=0
[2017/08/11 08:52:23] [debug] [router] input=tail.0 'DYNAMIC TAG'
[2017/08/11 08:52:23] [debug] [input tail.0] [mem buf] size = 26
[2017/08/11 08:52:23] [debug] [in_tail] file=/tmp/test.log read=15 lines=1
[2017/08/11 08:52:23] [debug] [in_tail] file=/tmp/test.log promote to TAIL_EVENT
[2017/08/11 08:52:28] [debug] [task->buffer] worker_id=0
[2017/08/11 08:52:28] [debug] [task] created task=0x7fd3d58330e0 id=0 OK
[2017/08/11 08:52:48] [error] [http_client] broken connection to 192.168.0.89:5044 ?
[2017/08/11 08:52:48] [error] [out_http] could not flush records (http_do=-1)
[2017/08/11 08:52:48] [debug] [retry] new retry created for task_id=0 attemps=1
[2017/08/11 08:52:48] [debug] [sched] retry=0x7fd3d58290f0 0 in 2 seconds
[2017/08/11 08:53:10] [error] [http_client] broken connection to 192.168.0.89:5044 ?
[2017/08/11 08:53:10] [error] [out_http] could not flush records (http_do=-1)
[2017/08/11 08:53:10] [debug] [retry] re-using retry for task_id=0 attemps=2
[2017/08/11 08:53:10] [debug] [sched] retry=0x7fd3d58290f0 0 in 8 seconds
[2017/08/11 08:53:38] [error] [http_client] broken connection to 192.168.0.89:5044 ?
[2017/08/11 08:53:38] [error] [out_http] could not flush records (http_do=-1)
[2017/08/11 08:53:38] [debug] [task] task_id=0 reached retry-attemps limit 2/1
[2017/08/11 08:53:38] [ warn] [engine] Task cannot be retried: task_id=0 thread_id=2 output=http.0
[2017/08/11 08:53:38] [debug] [task] destroy task=0x7fd3d58330e0 (task_id=0)
[2017/08/11 08:53:38] [debug] [dyntag tail.0] 0x7fd3d5872060 destroy (tag=test)
```

curl test:
```
curl -i -H  "Content-Type: application/json" -X POST -d '{"date":1502372062, "log":"curl_test"}' http://192.168.0.89:5044/
HTTP/1.1 200 OK
Content-Type: text/plain
Content-Length: 2

ok/
```

logtash log:
```
{
          "date" => 1502441543,
       "headers" => {
          "content_type" => "application/json",
          "request_path" => "/",
          "http_version" => "HTTP/1.1",
        "request_method" => "POST",
             "http_host" => "192.168.0.89:5044",
           "request_uri" => "/",
        "content_length" => "45"
    },
    "@timestamp" => 2017-08-11T08:52:28.006Z,
           "log" => "fluentbit_test",
      "@version" => "1",
          "host" => "172.17.0.3"
}
{
          "date" => 1502441543,
       "headers" => {
          "content_type" => "application/json",
          "request_path" => "/",
          "http_version" => "HTTP/1.1",
        "request_method" => "POST",
             "http_host" => "192.168.0.89:5044",
           "request_uri" => "/",
        "content_length" => "45"
    },
    "@timestamp" => 2017-08-11T08:52:50.002Z,
           "log" => "fluentbit_test",
      "@version" => "1",
          "host" => "172.17.0.3"
}
{
          "date" => 1502441543,
       "headers" => {
          "content_type" => "application/json",
          "request_path" => "/",
          "http_version" => "HTTP/1.1",
        "request_method" => "POST",
             "http_host" => "192.168.0.89:5044",
           "request_uri" => "/",
        "content_length" => "45"
    },
    "@timestamp" => 2017-08-11T08:53:18.003Z,
           "log" => "fluentbit_test",
      "@version" => "1",
          "host" => "172.17.0.3"
}
```

**Comment by edsiper** [comment]
@2solt 

__disclaimer__: this plugin was not designed to work with logstash

but let's see what the problem is...

```
[http_client] broken connection to 192.168.0.89:5044 ?
[2017/08/11 08:52:48] [error] [out_http] could not flush records (http_do=-1)
```

looks like the TCP connection was broken, I will do some tests and get back to you during the week.

**Comment by 2solt** [comment]
I amused that not design for it.
I was just reading fluent-bit's http output spec and logstash's http input spec, and I tough I give it a try.
Thanks for having a look tough.

**Comment by gibban69** [comment]
We see the same problem in a preproduction environment. Isn't it HTTP only...? Why would this be related to fluentd-bit <-> logstash compatibility?
Is there some configuration, like increasing the socket timeout, in the fluentdbit (td-agent-bit) output plugin that solves the problem?
If this plugin is not designed to work with logstash, which plugin would you recommend to use?

**Comment by seanmcguiganjcd** [comment]
Also seeing this behaviour connecting to elasticsearch.
`[2017/11/13 15:20:20] [error] [http_client] broken connection to logging.develop.farm ?
[2017/11/13 15:20:20] [ warn] [out_es] http_do=-1`

**Comment by edsiper** [comment]
when you face a "broken connection to logging.develop.farm" means the remote HTTP server closes the TCP connection, so the output plugin in Fluent Bit side trap the networking problem and issue a retry.

**Comment by gibban69** [comment]
I want to add that we see the problem on a localhost connection. That is on transfers from td-agent-bit to logstash http input on localhost. The setup is like this:
td-agent-bit, systemd input plugin, collects journald logs and forwards them, with forward plugin, from several hosts to this central td-agent-bit that fowards them to a local logstash http input.
Data rate is quite high, and errors logs mentioned from td-agent-bit are frequent. 
Is a new session created for each flush. Would a tcpdump file for wireshark help investigation?

**Comment by wkruse** [comment]
I am also getting a bunch of those running`fluent/fluent-bit:0.12.10` in Kubernetes cluster (setup like in http://fluentbit.io/documentation/0.12/installation/kubernetes.html, just replaced Elasticsearch Output with HTTP Output) and shipping logs with HTTP output to Logstash. I do see the logs arriving in Logstash though.

```
[2017/12/19 10:41:34] [error] [out_http] no upstream connections available
[2017/12/19 10:42:00] [error] [http_client] broken connection to matchbox.foo:9080 ?
[2017/12/19 10:42:00] [error] [out_http] could not flush records (http_do=-1)
```

**Comment by wkruse** [comment]
It seems there is a new connection for every HTTP POST and there seems to be always TCP Retransmissions. Could it be, that those TCP Retransmissions are causing the errors above?

```
 1 16:10:56.044048	172.17.4.11	172.17.4.253	TCP	74	39510 → 9080 [SYN] Seq=0 Win=29200 Len=0 MSS=1460 SACK_PERM=1 TSval=2623070665 TSecr=0 WS=1024
 2 16:10:56.044771	172.17.4.11	172.17.4.253	TCP	66	39510 → 9080 [ACK] Seq=1 Ack=1 Win=29696 Len=0 TSval=2623070666 TSecr=297768852
 4 16:10:56.048348	172.17.4.11	172.17.4.253	TCP	164	39510 → 9080 [PSH, ACK] Seq=1 Ack=1 Win=29696 Len=98 TSval=2623070669 TSecr=297768852 [TCP segment of a reassembled PDU]
 5 16:10:56.048356	172.17.4.11	172.17.4.253	HTTP	2166	POST / HTTP/1.1  (application/json)
 9 16:10:56.059407	172.17.4.11	172.17.4.253	TCP	718	[TCP Retransmission] 39510 → 9080 [PSH, ACK] Seq=1547 Ack=1 Win=29696 Len=652 TSval=2623070681 TSecr=297768856
10 16:10:56.060051	172.17.4.11	172.17.4.253	TCP	1514	[TCP Retransmission] 39510 → 9080 [ACK] Seq=99 Ack=1 Win=29696 Len=1448 TSval=2623070682 TSecr=297768867
11 16:10:56.131793	172.17.4.11	172.17.4.253	TCP	66	39510 → 9080 [ACK] Seq=2199 Ack=65 Win=29696 Len=0 TSval=2623070753 TSecr=297768938
12 16:10:56.132719	172.17.4.11	172.17.4.253	TCP	66	39510 → 9080 [ACK] Seq=2199 Ack=67 Win=29696 Len=0 TSval=2623070754 TSecr=297768939
23 16:11:16.335252	172.17.4.11	172.17.4.253	TCP	66	39510 → 9080 [FIN, ACK] Seq=2199 Ack=68 Win=29696 Len=0 TSval=2623090957 TSecr=297789142
```

**Comment by wkruse** [comment]
After digging a bit deeper with `Log_Level debug` and `tcpdump`,  it looks like`fluent-bit` doesn't accept the response from Logstash HTTP input.

Tailing the logs and getting metadata from the Kubernetes API server works.

```
[2017/12/20 19:17:19] [debug] [in_tail] removed /var/log/containers/kubernetes-dashboard-6584579d65-jb878_kube-system_kubernetes-dashboard-25806ffceea19f18c86b690b73ff2753edc720d5a2b8a14115d11dd882eeaa12.log
[2017/12/20 19:17:25] [debug] [in_tail] append new file: /var/log/containers/kubernetes-dashboard-6584579d65-wpr8l_kube-system_kubernetes-dashboard-3403313a196870409658cdbba3cfac58b15ef67fe5ef7f2eac30cb4cb0eb5519.log
[2017/12/20 19:17:26] [debug] [in_tail] add to scan queue /var/log/containers/kubernetes-dashboard-6584579d65-wpr8l_kube-system_kubernetes-dashboard-3403313a196870409658cdbba3cfac58b15ef67fe5ef7f2eac30cb4cb0eb5519.log, offset=0
[2017/12/20 19:17:26] [debug] [filter_kube] API Server (ns=kube-system, pod=kubernetes-dashboard-6584579d65-wpr8l) http_do=0, HTTP Status: 200
[2017/12/20 19:17:26] [debug] [input tail.0] [mem buf] size = 10163
[2017/12/20 19:17:26] [debug] [in_tail] file=/var/log/containers/kubernetes-dashboard-6584579d65-wpr8l_kube-system_kubernetes-dashboard-3403313a196870409658cdbba3cfac58b15ef67fe5ef7f2eac30cb4cb0eb5519.log read=1679 lines=12
[2017/12/20 19:17:26] [debug] [in_tail] file=/var/log/containers/kubernetes-dashboard-6584579d65-wpr8l_kube-system_kubernetes-dashboard-3403313a196870409658cdbba3cfac58b15ef67fe5ef7f2eac30cb4cb0eb5519.log promote to TAIL_EVENT
[2017/12/20 19:17:27] [debug] [task->buffer] worker_id=0
[2017/12/20 19:17:27] [debug] [task] created task=0x7f7e869841c0 id=2 OK
```

Sending the logs to Logstash works for **every** attempt, but `fluent-bit` thinks it **doesn't**.

```
[2017/12/20 19:17:47] [error] [http_client] broken connection to matchbox.foo:9080 ?
[2017/12/20 19:17:47] [error] [out_http] could not flush records (http_do=-1)
[2017/12/20 19:17:47] [debug] [retry] new retry created for task_id=2 attemps=1
[2017/12/20 19:17:47] [debug] [sched] retry=0x7f7e86817348 2 in 7 seconds

[2017/12/20 19:18:14] [error] [http_client] broken connection to matchbox.foo:9080 ?
[2017/12/20 19:18:14] [error] [out_http] could not flush records (http_do=-1)
[2017/12/20 19:18:14] [debug] [retry] re-using retry for task_id=2 attemps=2
[2017/12/20 19:18:14] [debug] [sched] retry=0x7f7e86817348 2 in 9 seconds

[2017/12/20 19:18:43] [debug] [task] task_id=2 reached retry-attemps limit 2/1
[2017/12/20 19:18:43] [ warn] [engine] Task cannot be retried: task_id=2 thread_id=2 output=http.0
[2017/12/20 19:18:43] [debug] [task] destroy task=0x7f7e869841c0 (task_id=2)
[2017/12/20 19:18:43] [debug] [dyntag tail.0] 0x7f7e86867240 destroy (tag=kube.var.log.containers.kubernetes-dashboard-6584579d65-wpr8l_kube-system_kubernetes-dashboard-3403313a196870409658cdbba3cfac58b15ef67fe5ef7f2eac30cb4cb0eb5519.log)
```

The corresponding HTTP POSTs and responses for **every** attempt.

```
29	20:17:28.044226	172.17.4.11	172.17.4.253	HTTP	4825	POST / HTTP/1.1  (application/json)
51	20:17:28.701586	172.17.4.253	172.17.4.11	HTTP	68	HTTP/1.1 200 OK  (text/plain)

90	20:17:55.042923	172.17.4.11	172.17.4.253	HTTP	4825	POST / HTTP/1.1  (application/json)
109	20:17:55.700270	172.17.4.253	172.17.4.11	HTTP	68	HTTP/1.1 200 OK  (text/plain)

137	20:18:24.042731	172.17.4.11	172.17.4.253	HTTP	4825	POST / HTTP/1.1  (application/json)
156	20:18:24.683120	172.17.4.253	172.17.4.11	HTTP	68	HTTP/1.1 200 OK  (text/plain)
```

The Logstash HTTP input plugin response.

```
HTTP/1.1 200 OK
Content-Type: text/plain
Content-Length: 2

ok
```

**Comment by edsiper** [comment]
so it's time to download logstash :)

work in process.

**Comment by edsiper** [comment]
thanks everybody for testing and troubleshooting this issue.

The root cause of the problem was the following: logstash send the response back in two rounds: 1. headers 2. body/payload. The current http_client was not processing/marking the payload on this specific condition leading to have the connection open waiting for more data.

The problem has been fixed by https://github.com/fluent/fluent-bit/commit/bf7895bd0257887837bbc6735382250f0f6000d6

The changes are available in 0.12 branch. If you can help to test the changes against different HTTP end-points in addition to logstash would be appreciated.

**Comment by edsiper** [comment]
Fluent Bit v0.12.11 has been released: http://fluentbit.io/announcements/v0.12.11/
